### PR TITLE
feat: LT-1190 End User Documents

### PR DESCRIPTION
**Pull request by markm5A8Z9F43913A**

### CODE REVIEW

File: ui_automation_tests/conftest.py [hunk 1]
```diff
@@ -286,6 +286,12 @@ def raise_clc_query(driver, control_code, description):
     exporter_hub.click_save_and_continue()
 
 
+@when("I click yes on the confirmation")
+def click_yes_on_confirm(driver):
+    shared = Shared(driver)
+    shared.click_confirm_yes()
```
Comment (phlptt) [inline ui_automation_tests/conftest.py]: can get this on one line 
Shared(driver).click_confirm_yes()
optional

File: ui_automation_tests/pages/application_overview_page.py [hunk 2]
```diff
@@ -42,6 +47,7 @@ def get_text_of_countries_selected(self):
         return self.driver.find_element_by_css_selector(self.show_countries_link).text
 
     def click_on_countries_selected(self):
+        self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight)")
```
Comment (phlptt) [inline ui_automation_tests/pages/application_overview_page.py]: i think there's a function in helpers.helpers for this , if not would be useful to add it there

File: ui_automation_tests/pages/application_overview_page.py [hunk 3]
```diff
@@ -55,3 +61,14 @@ def get_text_of_country_modal_content(self):
 
     def get_text_of_end_user_table(self):
         return self.driver.find_elements_by_css_selector(self.gov_tables)[len(self.driver.find_elements_by_css_selector(self.gov_tables))-1].text
+
+    def check_submit_is_enabled(self):
+        return self.driver.find_element_by_css_selector(self.submit_application_button).is_enabled()
+
+    def click_on_add_end_user_document(self):
+        self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight)")
```
Comment (phlptt) [inline ui_automation_tests/pages/application_overview_page.py]: same here

File: ui_automation_tests/pages/application_overview_page.py [hunk 4]
```diff
@@ -55,3 +61,14 @@ def get_text_of_country_modal_content(self):
...
+        self.driver.find_element_by_id(self.attach_end_user_document_link).click()
+
+    def click_delete_end_user_document(self):
+        self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight)")
```
Comment (phlptt) [inline ui_automation_tests/pages/application_overview_page.py]: same here

File: ui_automation_tests/pages/new_site_page.py [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import time
```
Comment (phlptt) [inline ui_automation_tests/pages/new_site_page.py]: can remove this import

File: ui_automation_tests/conftest.py [hunk 6]
```diff
@@ -286,6 +286,11 @@ def raise_clc_query(driver, control_code, description):
     exporter_hub.click_save_and_continue()
 
 
+@when("I click yes on the confirmation")
```
Comment (phlptt) [inline ui_automation_tests/conftest.py]: doesnt look like this is used

File: ui_automation_tests/pages/shared.py [hunk 7]
```diff
@@ -3,16 +3,28 @@ class Shared:
...
+    def get_text_of_error_messages(self):
+        return self.driver.find_element_by_css_selector(self.error_messages).text
 
     def click_continue(self):
         self.driver.find_element_by_css_selector(self.submit_button).click()
 
+    def click_back_link(self):
+        self.driver.find_element_by_css_selector(self.back_link).click()
+
+    def click_confirm_yes(self):
```
Comment (phlptt) [inline ui_automation_tests/pages/shared.py]: doesnt look like these are used click_confirm_yes and click_confirm_no

File: ui_automation_tests/step_defs/test_open_application.py [hunk 8]
```diff
@@ -14,17 +14,17 @@
...
 
 @then('I see good types error messages')
 def goods_type_errors(driver):
     shared = Shared(driver)
-    assert "This field may not be blank." in shared.get_text_of_error_message()
-    assert "This field is required." in shared.get_text_of_error_message(1)
-    assert "This field is required." in shared.get_text_of_error_message(2)
+    assert "This field may not be blank." in shared.get_text_of_error_messages()
+    assert "This field is required." in shared.get_text_of_error_messages()
+    assert "This field is required." in shared.get_text_of_error_messages()
```
Comment (phlptt) [inline ui_automation_tests/step_defs/test_open_application.py]: dont need this line

File: apply_for_a_licence/forms/end_user.py [hunk 9]
```diff
@@ -44,3 +47,20 @@ def new_end_user_form():
              pk='4',
              default_button_name='Save and continue')
     ])
+
+
+def attach_document_form(draft_url):
+    return Form(get_string('end_user.documents.attach_documents.title'),
+                get_string('end_user.documents.attach_documents.description'),
+                [
+                    FileUpload('documents')
```
Comment (5pence) [inline apply_for_a_licence/forms/end_user.py]: move [ FileUpload('documents') ], on line 54

File: apply_for_a_licence/services.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
+def add_document_data(request):
+    files = request.FILES.getlist("file")
+    if len(files) is 0:
```
Comment (5pence) [inline apply_for_a_licence/services.py]: if files: 
if the length is zero then it will return falsely?

File: conf/constants.py [hunk 11]
```diff
@@ -6,6 +6,7 @@
 GOODS_URL = '/goods/'
 CLCS_URL = APPLICATIONS_URL + 'clcs/'
 DOCUMENTS_URL = '/documents/'
+END_USER_DOCUMENTS_URL = '/end-user/document/'
```
Comment (5pence) [inline conf/constants.py]: in keeping with the constant and convention this endpoint ought to be '/end-user/documents/'  (plural)

File: drafts/services.py [hunk 12]
```diff
@@ -89,3 +89,22 @@ def get_draft_countries(request, pk):
...
+
+def post_draft_end_user_document(request, pk, json):
+    data = post(request, DRAFTS_URL + pk + END_USER_DOCUMENTS_URL, json)
+    return data.json(), data.status_code
+
+def delete_draft_end_user_documents(request, pk):
+    data = delete(request, DRAFTS_URL + pk + END_USER_DOCUMENTS_URL)
+    return data.status_code
+
+# def delete_good_document(request, pk, doc_pk):
```
Comment (5pence) [inline drafts/services.py]: if not using this method then remove

File: Pipfile [hunk 13]
```diff
@@ -21,6 +21,7 @@ python-json-logger = "==0.1.11"
 requests = "==2.21.0"
 stringcase = "==1.2.0"
 psycopg2 = "==2.7.7"
+psycopg2-binary = "==2.7.7"
```
Comment (efahy) [inline Pipfile]: Why is this required?

Comment (efahy) [reply Pipfile]: Just wonderin'

Comment (efahy) [reply Pipfile]: Chris S says remove this as it causes problems for people using docker

File: apply_for_a_licence/services.py (hunk 10 above)
Comment (efahy) [reply apply_for_a_licence/services.py]: Furthermore, apparently when comparing values in Python (as opposed to asserting that two variables refer to the same object), we should use `==` not `is`.  Please reflect throughout

File: apply_for_a_licence/views/common.py [hunk 14]
```diff
@@ -51,6 +59,14 @@ def get(self, request, **kwargs):
...
         external_locations, status_code = get_external_locations_on_draft(request, draft_id)
         ultimate_end_users, status_code = get_ultimate_end_users(request, draft_id)
+        end_user = data.get('draft').get('end_user')
+        if end_user:
+            draft_end_user_document, status_code = get_draft_end_user_documents(request, draft_id)
+            draft_end_user_document = draft_end_user_document.get('document')
+            can_submit = True if draft_end_user_document and draft_end_user_document['safe'] else False
+        else:
+            draft_end_user_document = None
+            can_submit = data.get('draft').get('licence_type') == 'open_licence'
```
Comment (efahy) [inline apply_for_a_licence/views/common.py]: Please define an enum for the possible values of `license_type` and use throughout the project

File: apply_for_a_licence/views/common.py [hunk 15]
```diff
@@ -346,3 +368,85 @@ def get(self, request, **kwargs):
...
+        draft, status_code = get_draft(request, draft_id)
+
+        if draft.get('draft').get('licence_type') == 'standard_licence':
+            form = attach_document_form(reverse('apply_for_a_licence:overview', kwargs={'pk': draft_id}))
+
+            return form_page(request, form, extra_data={'draft_id': draft_id})
+        else:
+            return redirect(reverse_lazy('apply_for_a_licence:overview', kwargs={'pk': draft_id}))
+
+    @csrf_exempt
```
Comment (efahy) [inline apply_for_a_licence/views/common.py]: Do you need this, as well as the `@method_decorator(csrf_exempt, 'dispatch')` line above?

File: apply_for_a_licence/views/common.py [hunk 16]
```diff
@@ -346,3 +368,85 @@ def get(self, request, **kwargs):
         }
 
         return render(request, 'apply_for_a_licence/ultimate_end_users/index.html', context)
+
+
+@method_decorator(csrf_exempt, 'dispatch')
+class AttachDocuments(TemplateView):
+    def get(self, request, **kwargs):
+        draft_id = str(kwargs['pk'])
+        draft, status_code = get_draft(request, draft_id)
```
Comment (efahy) [inline apply_for_a_licence/views/common.py]: I think we should check the status code before proceeding

File: apply_for_a_licence/views/common.py [hunk 17]
```diff
@@ -346,3 +368,85 @@ def get(self, request, **kwargs):
...
+        draft_id = str(kwargs['pk'])
+        data, error = add_document_data(request)
+
+        if error:
+            return error_page(None, 'We had an issue uploading your files. Try again later.')
+
+        # Send LITE API the file information
+        draft_end_user_documents, status_code = post_draft_end_user_document(request, draft_id, data)
+
+        if 'errors' in draft_end_user_documents:
```
Comment (efahy) [inline apply_for_a_licence/views/common.py]: Maybe also include a check of the status code?

File: apply_for_a_licence/views/common.py [hunk 18]
```diff
@@ -346,3 +368,85 @@ def get(self, request, **kwargs):
...
+        if document['safe']:
+            original_file_name = document['name']
+
+            # Stream file
+            def generate_file(result):
+                for chunk in iter(lambda: result['Body'].read(STREAMING_CHUNK_SIZE), b''):
+                    yield chunk
+
+            s3 = s3_client()
+            s3_response = s3.get_object(Bucket=AWS_STORAGE_BUCKET_NAME, Key=document['s3_key'])
```
Comment (efahy) [inline apply_for_a_licence/views/common.py]: A lot of this code is very similar to code in `goods/views.py` - can we employ a re-usable method?

File: apply_for_a_licence/views/common.py (hunk 15 above)
Comment (BenAndrew-Kainos) [reply apply_for_a_licence/views/common.py]: Yes the @method_decorator is needed to allow @csrf_exempt to be used on methods inside the view class. @csrf_exempt is needed to allow file streaming.

File: apply_for_a_licence/views/common.py [hunk 19]
```diff
@@ -346,3 +367,74 @@ def get(self, request, **kwargs):
...
+
+    @csrf_exempt
+    def post(self, request, **kwargs):
+        self.request.upload_handlers.insert(0, S3FileUploadHandler(request))
+
+        draft_id = str(kwargs['pk'])
+        data, error = add_document_data(request)
+
+        if error:
+            return error_page(None, 'We had an issue uploading your files. Try again later.')
```
Comment (lukesa) [inline apply_for_a_licence/views/common.py]: should we use a string from content repo?

File: apply_for_a_licence/views/common.py [hunk 20]
```diff
@@ -346,3 +367,74 @@ def get(self, request, **kwargs):
...
+
+        # Send LITE API the file information
+        end_user_document, status_code = post_end_user_document(request, draft_id, data)
+
+        if status_code != 201:
+            return error_page(None, 'We had an issue uploading your files. Try again later.')
```
Comment (lukesa) [inline apply_for_a_licence/views/common.py]: should we use a string from content repo?

File: apply_for_a_licence/views/common.py [hunk 21]
```diff
@@ -346,3 +367,74 @@ def get(self, request, **kwargs):
...
+        draft_id = str(kwargs['pk'])
+
+        documents, status_code = get_end_user_document(request, draft_id)
+
+        document = documents['document']
+
+        if document['safe']:
+            return download_document_from_s3(document['s3_key'], document['name'])
+        else:
+            return error_page(None, 'We had an issue downloading your file. Try again later.')
```
Comment (lukesa) [inline apply_for_a_licence/views/common.py]: should we use a string from content repo?

File: apply_for_a_licence/views/common.py [hunk 22]
```diff
@@ -346,3 +367,74 @@ def get(self, request, **kwargs):
...
+
+    def post(self, request, **kwargs):
+        draft_id = str(kwargs['pk'])
+        option = request.POST.get('delete_document_confirmation')
+        if option is None:
+            return redirect(reverse('apply_for_a_licence:delete_document', kwargs={'pk': draft_id}))
+        elif option == 'yes':
+            status_code = delete_end_user_document(request, draft_id)
+            if status_code is not 204:
+                return error_page(None, 'We had an issue deleting your files. Try again later.')
```
Comment (lukesa) [inline apply_for_a_licence/views/common.py]: should we use a string from content repo?
